pallets/avn-oracle/src/lib.rs

```diff
@@ -0,0 +1,569 @@
+            match call {
+                Call::submit_price { rates, submitter, signature } =>
+                    if AVN::<T>::signature_is_valid(
+                        &(PRICE_SUBMISSION_CONTEXT, rates, Nonce::<T>::get()),
+                        &submitter,
+                        signature,
+                    ) {
+                        ValidTransaction::with_tag_prefix("SubmitAvtPrice")
+                            .and_provides(vec![(
+                                PRICE_SUBMISSION_CONTEXT,
+                                rates,
+                                Nonce::<T>::get(),
+                                submitter.account_id.clone(),
+                            )
+                                .encode()])
+                            .priority(TransactionPriority::max_value())
+                            .longevity(64_u64)
+                            .propagate(true)
+                            .build()
+                    } else {
+                        InvalidTransaction::Custom(1u8).into()
+                    },
+                Call::clear_consensus { submitter, signature } =>
+                    if AVN::<T>::signature_is_valid(
+                        &(CLEAR_CONSENSUS_SUBMISSION_CONTEXT, Nonce::<T>::get()),
+                        &submitter,
+                        signature,
+                    ) {
+                        ValidTransaction::with_tag_prefix("ClearConsensus")
+                            .and_provides(vec![(
+                                CLEAR_CONSENSUS_SUBMISSION_CONTEXT,
+                                Nonce::<T>::get(),
+                                submitter.account_id.clone(),
+                            )
+                                .encode()])
+                            .priority(TransactionPriority::max_value())
+                            .longevity(64_u64)
+                            .propagate(true)
+                            .build()
+                    } else {
+                        InvalidTransaction::Custom(1u8).into()
+                    },
+                _ => InvalidTransaction::Call.into(),
+            }
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        fn submit_price_if_required(
+            submitter: &Validator<T::AuthorityId, T::AccountId>,
+        ) -> Result<(), DispatchError> {
+            if !Self::should_query_rates() {
+                return Ok(());
+            }
+
+            let current_block = <frame_system::Pallet<T>>::block_number();
+            let last_submission_block = LastPriceSubmission::<T>::get();
+            let nonce = Nonce::<T>::get();
+
+            let guard_lock_name =
+                Self::create_guard_lock(b"submit_price::", nonce, &submitter.account_id);
+
+            if current_block >=
+                last_submission_block +
+                    BlockNumberFor::<T>::from(T::PriceRefreshRangeInBlocks::get())
```

Comment:
This could be abstracted in a small helper function. Its also safer to use `saturating` when doing arithmetics. Example:
```
fn is_refresh_due(current: BlockNumberFor<T>, last: BlockNumberFor<T>) -> bool {
        current >= last.saturating_add(BlockNumberFor::<T>::from(T::PriceRefreshRangeInBlocks::get()))
    }
```